addons/binding/org.openhab.binding.allplay/src/main/java/org/openhab/binding/allplay/handler/AllPlayHandler.java

```diff
@@ -0,0 +1,499 @@
+    }
+
+    private void handleVolumeCommand(Command command) throws SpeakerException {
+        if (command instanceof PercentType) {
+            speaker.volume().setVolume(convertPercentToAbsoluteVolume((PercentType) command));
+        } else if (command instanceof IncreaseDecreaseType) {
+            int stepSize = (command == IncreaseDecreaseType.DECREASE ? -getVolumeStepSize() : getVolumeStepSize());
+            speaker.volume().adjustVolume(stepSize);
+        }
+    }
+
+    @Override
+    public void onPlayStateChanged(PlayState playState) {
+        updatePlayState(playState);
+        updatePlaylistItemsState(playState.getPlaylistItems());
+    }
+
+    @Override
+    public void onPlaylistChanged() {
+        logger.debug(speaker.getName() + ": Playlist changed: No action");
+    }
+
+    @Override
+    public void onLoopModeChanged(LoopMode loopMode) {
+        logger.debug(speaker.getName() + ": LoopMode changed to " + loopMode);
+        updateState(LOOP_MODE, new StringType(loopMode.toString()));
+
+    }
+
+    @Override
+    public void onShuffleModeChanged(ShuffleMode shuffleMode) {
+        logger.debug(speaker.getName() + ": ShuffleMode changed to " + shuffleMode);
+        updateState(SHUFFLE_MODE, new StringType(shuffleMode.toString()));
+    }
+
+    @Override
+    public void onMuteChanged(boolean mute) {
+        logger.debug(speaker.getName() + ": Mute changed to " + mute);
+        updateState(MUTE, mute ? OnOffType.ON : OnOffType.OFF);
+    }
+
+    @Override
+    public void onVolumeChanged(int volume) {
+        logger.debug(speaker.getName() + ": Volume changed to " + volume);
+        try {
+            updateState(VOLUME, convertAbsoluteVolumeToPercent(volume));
+        } catch (SpeakerException e) {
+            logger.warn("Cannot convert new volume to percent", e);
+        }
+    }
+
+    @Override
+    public void onVolumeControlChanged(boolean enabled) {
+        updateState(VOLUME_CONTROL, enabled ? OnOffType.ON : OnOffType.OFF);
+    }
+
+    @Override
+    public void onZoneChanged(String zoneId, int timestamp, Map<String, Integer> slaves) {
+        logger.debug(speaker.getName() + ": Zone changed to " + zoneId);
+        updateState(ZONE_ID, new StringType(zoneId));
+    }
+
+    private void updatePlayState(PlayState playState) {
+        logger.debug(speaker.getName() + ": PlayState changed to " + playState);
+        updateState(PLAY_STATE, new StringType(playState.getState().toString()));
+
+        if (playState.getState() == State.PLAYING) {
+            updateState(CONTROL, PlayPauseType.PLAY);
+        } else {
+            updateState(CONTROL, PlayPauseType.PAUSE);
+        }
+    }
+
+    private void updatePlaylistItemsState(List<PlaylistItem> items) {
+        if (!items.isEmpty()) {
+            PlaylistItem currentItem = items.iterator().next();
+            updateCurrentItemState(currentItem);
+        } else {
+            updateState(CURRENT_ARTIST, UnDefType.UNDEF);
```

Comment:
It probably should rather be UnDefType.NULL since you know that it is not set (while UNDEF is rather for "I have no idea what it is or at least I cannot express it").